pySOT/strategy.py

```diff
@@ -850,3 +850,447 @@ def generate_evals(self, num_pts):
+        self.num_cand = num_cand
+
+        self.sampling_radius = 0.2
+        self.record_queue = []  # Completed records that haven't been processed
+        self.num_exp = exp_design.num_pts  # We need this later
+        self.ncenters = ncenters
+        self.evals = [] # List that stores all evaluated points as SOPRecord instances
+        self.centers = [] #List that stores all current center points is SOPCenter instances
+        self.F_ranked = None # The evaluated points stored as numpy array with info of rank and tabu status
+        self.d_thresh = 1.0
+
+        super().__init__(max_evals=max_evals, opt_prob=opt_prob,
+                         exp_design=exp_design, surrogate=surrogate,
+                         asynchronous=asynchronous, batch_size=batch_size,
+                         extra_points=extra_points, extra_vals=extra_vals,
+                         reset_surrogate=reset_surrogate)
+
+    def check_input(self):
+        """Check inputs."""
+        super().check_input()
+        if not isinstance(self.ncenters, int) and self.ncenters > 3:
+            raise ValueError("ncenters should be an integer greater than 3")
+        if not self.asynchronous:
+            if not self.ncenters >= self.batch_size:
+                raise ValueError("Batch size should be less than or equal to ncenters")
+
+    def on_initial_completed(self, record):
+        """Handle completed evaluation in initial phase"""
+        super().on_initial_completed(record)
+
+        srec = SopRecord(np.copy(record.params[0]),record.value,0,0,0,self.sampling_radius)
+        self.evals.append(srec)
+
+    def on_adapt_completed(self, record):
+        """Handle completed evaluation in phase 2."""
+        super().on_adapt_completed(record)
+        self.record_queue.append(record)
+
+        # Initiate a new SOP Record for new completed evaluation
+        center_index = None
+        srec = SopRecord(np.copy(record.params[0]),record.value,0,0,0,self.sampling_radius)
+        self.evals.append(srec)
+        ncenters = len(self.centers)
+        for i in range(ncenters): #Update the location of new point in its corresponding center
+            if np.array_equal(np.copy(record.params[0]),self.centers[i].new_point):
+                self.centers[i].add_index(self.num_evals - 1)
+                center_index = i
+                break
+
+        if self.asynchronous:  # Process immediately
+            self.adjust_memory(center_index)
+        elif (not self.batch_queue) and self.pending_evals == 0:  # Batch
+            self.adjust_memory()
+
+    def generate_evals(self, num_pts):
+        """Generate the next adaptive sample points."""
+
+        # Update the list of center points
+        if self.F_ranked is None: # If this is the first entry into adaptive phase
+            self.update_ranks()
+            self.create_center_list()
+        else:
+            self.update_center_list()
+
+        # Compute dycors perturbation probability
```

Comment:
I dont think a super call would work. Since, `SOPStrategy` generates candidates around multiple points. We could avoid code duplication by making a function that computes the DYCORS probability and call that function in both 'SOPStrategy' and `DYCORSStrategy`?

Reply:
I dont think a super call would work. Since, SOPStrategy generates candidates around multiple points. We could avoid code duplication by making a function that computes the DYCORS probability and call that function in both `SOPStrategy` and DYCORSStrategy?

Reply:
I dont think a super call would work. Since, `SOPStrategy` generates candidates around multiple points. We could avoid code duplication by making a function that computes the DYCORS probability and call that function in both 'SOPStrategy' and `DYCORSStrategy`?

Reply:
I dont think a super call would work. Since, `SOPStrategy` generates candidates around multiple points. We could avoid code duplication by making a function that computes the DYCORS probability and call that function in both `SOPStrategy` and `DYCORSStrategy`?